cirq/protocols/decompose.py

```diff
@@ -309,7 +309,10 @@ def decompose_once_with_qubits(val: Any,
 @overload
 def decompose_once_with_qubits(val: Any,
                                qubits: Iterable['cirq.QubitId'],
-                               default: TDefault,
+                               # NOTE: should be TDefault instead of Any, but
+                               # mypy has false positive errors when setting
+                               # default to None.
```

Comment:
Well, mypy may be right: TDefault is a type parameter and it may be set to a type which doesn't admit `None` value. If we do need to allow `None` then default should have type `Optional[TDefault]`. Am I missing something?

Reply:
If a generic type parameter only occurs for a single input, that input should be used to infer the value of the type parameter. So if I pass `None` for default, `TDefault` should become `NoneType` and if I pass `"text"` it should become `str`.